app/controllers/settings/ext_cert_controller.rb

```diff
@@ -0,0 +1,260 @@
+          pillar_model_key: :external_cert_velum_cert
+        },
+        key:  {
+          key_string:       get_val_from_form(:velum_key),
+          pillar_model_key: :external_cert_velum_key
+        }
+      },
+      kubeapi: {
+        name: KUBEAPI_NAME,
+        cert: {
+          cert_string:      get_val_from_form(:kubeapi_cert),
+          pillar_model_key: :external_cert_kubeapi_cert
+        },
+        key:  {
+          key_string:       get_val_from_form(:kubeapi_key),
+          pillar_model_key: :external_cert_kubeapi_key
+        }
+      },
+      dex:     {
+        name: DEX_NAME,
+        cert: {
+          cert_string:      get_val_from_form(:dex_cert),
+          pillar_model_key: :external_cert_dex_cert
+        },
+        key:  {
+          key_string:       get_val_from_form(:dex_key),
+          pillar_model_key: :external_cert_dex_key
+        }
+      }
+    }
+  end
+
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+  def upload_validate(key_cert_map)
+    # Do nothing if both cert/key are empty
+    if key_cert_map[:cert][:cert_string].empty? && key_cert_map[:key][:key_string].empty?
+      true
+    # Prevent upload unnless both cert/key are present
+    elsif key_cert_map[:cert][:cert_string].empty? || key_cert_map[:key][:key_string].empty?
+      message = "Error with #{key_cert_map[:name]}, certificate and key must be uploaded together."
+      render_failure_event(message)
+    # Validate cert/key and verify that they match
+    else
+      cert = read_cert(key_cert_map[:cert][:cert_string])
+      key = read_key(key_cert_map[:key][:key_string])
+
+      # Check certificate valid format
+      unless cert
+        message = "Invalid #{key_cert_map[:name]} certificate, check format and try again."
+        return render_failure_event(message)
+      end
+
+      # Check key valid format
+      unless key
+        message = "Invalid #{key_cert_map[:name]} key, check format and try again."
+        return render_failure_event(message)
+      end
+
+      # Check that key matches certificate
+      unless cert.check_private_key(key)
+        message = "#{key_cert_map[:name]} Certificate/Key pair invalid.  Ensure Certificate" \
+        " and Key are matching."
+        return render_failure_event(message)
+      end
+
+      # Check that cert has valid date range
+      return false unless cert_date_check(cert)
+
+      # Moved to another task
+      # Check that hostname is in SubjectAltName of cert
+      # return false unless hostname_check(key_cert_map[:name], cert)
+
+      # Moved to another task
+      # Check the trust chain is valid
+      # return false unless trust_chain_verify(cert)
+
+      # Everything's good!
+      true
+    end
+  end
+  # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def cert_parse(cert_string)
+    params = {}
+
+    # Check if certificate exists, assume that validation has already occured
+    if !cert_string
```

Comment:
Doesn't ruby style prefer unless instead of 'if !' ?

Reply:
Apparently this is only true with a hanging `if` statement.  If the `else` block is included, rucocop expects `if`.  This is the error message from rubocop if you use `unless` with `else`:
`Do not use unless with else. Rewrite these with the positive case first.`